runner/startup.sh

```diff
@@ -1,70 +1,75 @@
-}
+# This path can only be customized through arguments passed to the dockerd cli
+# at startup. Parsing the supervisor docker.conf file is cumbersome, and there
+# is little reason to allow users to customize this. Hence, we hardcode the path
+# here.
+#
+# See: https://docs.docker.com/engine/reference/commandline/dockerd/
+readonly dockerd_config_path=/etc/docker/daemon.json
+readonly dockerd_supervisor_config_path=/etc/supervisor/conf.d/dockerd.conf
 
-sudo /bin/bash <<SCRIPT
-mkdir -p /etc/docker
+# Extend existing config...
+dockerd_config=$(! cat $dockerd_config_path 2>&-)
 
-echo "{}" > /etc/docker/daemon.json
+# ...and fallback to an empty object if there is no existing configuration, or
+# otherwise the following `jq` merge commands would not produce anything.
+if [[ $dockerd_config == '' ]]; then
+  dockerd_config='{}'
+fi
 
-if [ -n "${MTU}" ]; then
-jq ".\"mtu\" = ${MTU}" /etc/docker/daemon.json > /tmp/.daemon.json && mv /tmp/.daemon.json /etc/docker/daemon.json
-# See https://docs.docker.com/engine/security/rootless/
-echo "environment=DOCKERD_ROOTLESS_ROOTLESSKIT_MTU=${MTU}" >> /etc/supervisor/conf.d/dockerd.conf
+readonly MTU=${MTU:-}
+if [[ $MTU ]]; then
+  # We have to verify that this value is a valid integer because we are going to
+  # start docker in the background and invalid values would never be surfaced.
+  # Well, except that startup fails and our users are sent on a nice debugging
+  # session.
+  if [[ $MTU =~ ^[1-9][0-9]*$ ]]; then
+    # See https://docs.docker.com/engine/security/rootless/
+    sudo tee -a $dockerd_supervisor_config_path 1>&- <<<"environment=DOCKERD_ROOTLESS_ROOTLESSKIT_MTU=$MTU"
+    dockerd_config=$(jq -S ".mtu = $MTU" <<<"$dockerd_config")
+  else
+    # shellcheck source=runner/logger.bash
+    source logger.bash
+    log.error "Docker MTU must be an integer, continuing bootstrapping without it, got: $MTU"
+    MTU=
+  fi
 fi
 
-if [ -n "${DOCKER_REGISTRY_MIRROR}" ]; then
-jq ".\"registry-mirrors\"[0] = \"${DOCKER_REGISTRY_MIRROR}\"" /etc/docker/daemon.json > /tmp/.daemon.json && mv /tmp/.daemon.json /etc/docker/daemon.json
+if [[ ${DOCKER_REGISTRY_MIRROR:-} != '' ]]; then
+  dockerd_config=$(jq -S --arg url "$DOCKER_REGISTRY_MIRROR" '."registry-mirrors" += ["$url"]' <<<"$dockerd_config")
 fi
-SCRIPT
 
-dump() {
-  local path=${1:?missing required <path> argument}
-  shift
-  printf -- "%s\n---\n" "${*//\{path\}/"$path"}" 1>&2
-  cat "$path" 1>&2
-  printf -- '---\n' 1>&2
-}
+sudo mkdir -p ${dockerd_config_path%/*}
+sudo tee $dockerd_config_path 1>&- <<<"$dockerd_config"
 
-for config in /etc/docker/daemon.json /etc/supervisor/conf.d/dockerd.conf; do
-  dump "$config" 'Using {path} with the following content:'
+for config in $dockerd_config_path $dockerd_supervisor_config_path; do
+  (
+    echo "Using '$config' with the following content <<CONFIG"
+    cat $config
+    echo 'CONFIG'
+  ) 1>&2
 done
 
-log.debug 'Starting supervisor daemon'
-sudo /usr/bin/supervisord -n >> /dev/null 2>&1 &
```

Comment:
We forcefully start supervisord in the foreground to then directly send it to the background again. We also set stdout to append to `/dev/null` and redirect its stderr to stdout. I really don't understand why we do what we do here. 😕